oshmem/shmem/c/shmem_mcs_lock.c

```diff
@@ -0,0 +1,300 @@
+	 * Get the previous tail PE.
+	 * int prev_tail	 = shmem_atomic_swap(tail, new_tail_req,
+				     mcs_tail_owner);
+	 */
+	rc = MCA_ATOMIC_CALL(swap(oshmem_ctx_default,
+			  (void *)tail,
+			  (void*)&prev_tail,
+			  OSHMEM_ATOMIC_PTR_2_INT(&new_tail_req,
+						  sizeof(new_tail_req)),
+			  sizeof(int),
+			  mcs_tail_owner));
+	RUNTIME_CHECK_RC(rc);
+	prev_tailpe	= GET_PE(prev_tail);
+
+	if (SIGNAL(prev_tail)) {
+		/** Someone else has got the lock before this PE
+		 *  Adding this PE to the previous tail PE's Next pointer
+		 *  Substract the SIGNAL Bit to avoid changing it.
+		 *  shmem_atomic_add(next, my_pe - NEXT_MASK,
+				     prev_tailpe);
+		 **/
+		tmp_val = my_pe - NEXT_MASK;
+	        rc = MCA_ATOMIC_CALL(add(oshmem_ctx_default,
+					 (void*)next, tmp_val,
+					 sizeof(int), prev_tailpe));
+		RUNTIME_CHECK_RC(rc);
+
+		/** This value to be changed eventually by predecessor
+		 *  when its lock is released.
+		 *  Need to be done atomically to avoid any races where
+		 *  next pointer is modified by another PE acquiring or
+		 *  releasing this.
+		 **/
+		rc = MCA_ATOMIC_CALL(add(oshmem_ctx_default,
+					 (void *)next, SIGNAL_MASK,
+					 sizeof(int), my_pe));
+		RUNTIME_CHECK_RC(rc);
+		MCA_SPML_CALL(quiet(oshmem_ctx_default));
+		/**
+		 * Wait for predecessor release lock to this PE
+		 * signal to false.
+		 * int curr = shmem_atomic_fetch(next, my_pe);
+		 */
+	        rc = MCA_ATOMIC_CALL(fadd(oshmem_ctx_default,
+					  (void*)next, (void*)&curr,
+					  tval, sizeof(int), my_pe));
+	        RUNTIME_CHECK_RC(rc);
+		while (SIGNAL(curr)) {
+		    rc = MCA_SPML_CALL(wait((void*)next, SHMEM_CMP_NE,
+					(void*)&curr, SHMEM_INT));
+		    RUNTIME_CHECK_RC(rc);
+		    rc = MCA_ATOMIC_CALL(fadd(oshmem_ctx_default, (void*)next,
+					      (void*)&curr, tval, sizeof(int),
+					      my_pe));
+		    RUNTIME_CHECK_RC(rc);
+		}
+	}
+	/** else.. this pe has got the lock as no one else had it */
+}
+
+static inline void
+shmem_mcs_internal_clear_lock(long *lockp)
+{
+	lock_t	*lock			= (lock_t *) lockp;
+	int	mcs_tail_owner		= MCSQ_TAIL_OWNER(lock);
+	int	*tail			= &(lock->tail);
+	int	*next			= &(lock->next);
+	int	my_pe			= shmem_my_pe();
+	int	next_value		= 0;
+	int	swap_cond		= 0;
+	int	prev_value		= 0;
+	int	tval			= 0;
+	int	val_tmp			= 0;
+	int	nmask			= 0;
+	int	a_val			= 0;
+	int	rc			= 0;
+
+	/**
+	 * Can make atomic fetch to be safe in non-cc architecutres
+	 * Has impact on performance
+	 *
+	 * int next_value = __atomic_load_n(next, __ATOMIC_SEQ_CST);
+	 * int next_value = shmem_atomic_fetch(next, my_pe);
+	 * shmem_quiet();
+	 */
+        rc = MCA_ATOMIC_CALL(fadd(oshmem_ctx_default,
+				  (void*)next, (void*)&next_value,
+				  tval, sizeof(int), my_pe));
+        RUNTIME_CHECK_RC(rc);
+	MCA_SPML_CALL(quiet(oshmem_ctx_default));
+
+	if (next_value == NEXT_MASK) {
+		swap_cond = SET_SIGNAL(my_pe);
+
+		/**
+		 * prev_value = shmem_atomic_compare_swap(tail, swap_cond, 0,
+		 *					   mcs_tail_owner);
```

Comment:
isn't it almost the same what is used below?
same comment is relevant for other similar comments in this file

Reply:
I have removed this comment.